### PR TITLE
map: support larger bpf_map definitions

### PR DESCRIPTION
**Pull request by higebu**
* Add TestLoadInvalidMap to elf_test
* Add invalid_map to testdata

### CODE REVIEW

**Comment by lmb** [comment]
Hi higebu,

Thanks for your contribution! It seems like this PR reads numa_node, but doesn't do anything with it?
For it to be useful we'd have to plumb it through to MapSpec and NewMap somehow.

Can you describe your use case for NUMA local maps a bit? So far I haven't added support since I don't know what the best API for it would be.

**Comment by higebu** [comment]
@lmb Thank you for your comment.
My initial goal is to fix "unknown fields in definition" error with ebp-dump on Linux 4.18. You can see the error with objects in samples/bpf like xdp1_kern.o.

```
linux/samples/bpf$ ebpf-dump xdp1_kern.o
panic: load maps: map rxcnt: unknown fields in definition

goroutine 1 [running]:
main.main()
        /home/yuya/go/src/github.com/newtools/ebpf/cmd/ebpf-dump/main.go:27 +0x86f
```

But I added NumaNode to MapSpec and TestMapNumaNode to map_test for numa_node support.

**Comment by lmb** [comment]
Thanks for updating the PR!

I've looked at the kernel bpf_map struct, and several things come to my mind:

- I think numa_node shouldn't be part of bpf_map in the first place. That's 
because it's impossible to know at compile time what the layout of the 
target machine is. Point in case, I couldn't find any examples in the 
kernel where numa_node is being set.
- On the other hand, it's probably wise to be compatible with the kernel / 
libbpf bpf_map definition (there are competing ones, e.g. in iproute2). 
Maybe we read numa_node in elf.go, but throw an error if it's non zero. 
That would mean no plumbing to MapSpec for now (that would come once we 
figure out a use case for numa local maps).

These are somewhat involved, so maybe we can for now emulate what the 
kernel does: accept larger bpf_map definitions, as long as the unknown part 
is all zeroes. This would fix dumping the kernel structs and give some 
forward compatibility.

What do you think?
-

**Comment by higebu** [comment]
@lmb It makes sense to me. I fixed this PR for support larger bpf_map definitions. Please review it.
Thank you.

File: elf.go [hunk 1]
```diff
@@ -215,8 +215,9 @@ func (ec *elfCode) loadMaps(mapSections map[int]*elf.Section) (map[string]*MapSp
 				return nil, errors.Errorf("map %v: can't read inner map index", name)
 			}
 
-			if rd.Len() != 0 {
-				return nil, errors.Errorf("map %v: unknown fields in definition", name)
+			var unknown uint32
+			if rd.Len() > 0 && binary.Read(rd, ec.ByteOrder, &unknown) != nil || unknown != 0 {
```
Comment (lmb) [inline elf.go]: Suggestion: use something like `bytes.Count(rd.Bytes(), []byte{0}) != rd.Len()`. Then we handle arbitrary zero tails.

**Comment by lmb** [comment]
Looks good, thank you for contributing!

**Pull request merged**